Makefile

```diff
@@ -54,25 +63,46 @@ fmt: ## Run go fmt against code.
 vet: ## Run go vet against code.
 	go vet ./...
 
-.PHONY: test
-test: manifests generate fmt vet envtest ## Run tests.
+.PHONY: test test-e2e e2e kind-load kind-cluster kind-cluster-cleanup
+test: manifests generate fmt vet envtest test-e2e ## Run all tests.
 	KUBEBUILDER_ASSETS="$(shell $(ENVTEST) use $(ENVTEST_K8S_VERSION) --bin-dir $(LOCALBIN) -p path)" go test ./... -coverprofile cover.out
 
+FOCUS := $(if $(TEST),-v -focus "$(TEST)")
+E2E_FLAGS ?= ""
+test-e2e: ginkgo ## Run the e2e tests
+	$(GINKGO) --tags $(GO_BUILD_TAGS) $(E2E_FLAGS) -trace -progress $(FOCUS) test/e2e
+
+e2e: KIND_CLUSTER_NAME=operator-controller-e2e
+e2e: run test-e2e kind-cluster-cleanup ## Run e2e test suite on local kind cluster
+
+kind-load: kind ## Loads the currently constructed image onto the cluster
+	$(KIND) load docker-image $(IMG) --name $(KIND_CLUSTER_NAME)
+
+kind-cluster: kind kind-cluster-cleanup ## Standup a kind cluster
+	$(KIND) create cluster --name ${KIND_CLUSTER_NAME}
+	$(KIND) export kubeconfig --name ${KIND_CLUSTER_NAME}
+
+kind-cluster-cleanup: kind ## Delete the kind cluster
+	$(KIND) delete cluster --name ${KIND_CLUSTER_NAME}
+
 ##@ Build
 
 .PHONY: build
 build: manifests generate fmt vet ## Build manager binary.
 	go build -o bin/manager main.go
 
 .PHONY: run
-run: manifests generate fmt vet ## Run a controller from your host.
-	go run ./main.go
+run: docker-build kind-cluster kind-load install deploy wait ## Build the operator-controller then deploy it into a new kind cluster.
```

Comment:
Do we really want to change the behavior of the `run` target? Would another target make sense?

Also a bit concerned on ordering here, if `make` is run with `-j`. (I not sure if order-only prerequisites will make it any better, but the prerequisites of all these prerequisites need to be proper.)

Reply:
A parallel build will definitely change the order here.
Might want to add:
```
.NOTPARALLEL: run

Reply:
This change was also motivated by consistency with rukpak, where `make run` does exactly the same thing. I could add a `run-local` perhaps if we want to keep this capability?
For the `.NOTPARALLEL` thing, I could try adding a line in the makefile that covers all the commands with dependencies, since nearly all of them (not just `run`) would be broken with `-j`.

Reply:
Updated, I added `.NOTPARALLEL:` which disables it for all targets. They all pretty much all need to be run sequentially so I think this is fine.